### PR TITLE
[misc] feat: Add Cambricon MLU support for VeOmni

### PR DESCRIPTION
**Pull request by sc1915**
### What does this PR do?

> This pull request aims to provide support for the MLU backend.

### Test

> There is a example training **Qwen3-VL-30B-A3B-Instruct** using fused_moe_kernel `fused_mlu`: 
```
bash train.sh tasks/train_vlm.py configs/multimodal/qwen3_vl/qwen3_vl_moe.yaml --train.accelerator.ep_size 8
```
Training loss:
image

moe_implementation="fused_mlu" in mlu, and moe_implementation="fused_triton" in gpu

### Design & Code Changes

- Building on the existing hardware infrastructure, this PR introduces MLU backend support, starting with the `fused_mlu` kernel (fused_moe_kernel in ``veomni/ops/kernels/moe/mlu_group_gemm.py``). More MLU fused kernels will be added over time.

- `veomni/ops/kernel_registry.py` allows `device_type` accept list or str.

## New Features
* Added Cambricon MLU support for model initialization, distributed training, profiling, device management, and accelerator detection.
* Added MLU-compatible fused MoE, grouped GEMM, normalization, and related kernel operations.
* Added MLU device detection and visible-device configuration during training.
* Added selectable `mlu` and `mlu_triton` MoE backends, including validation of required hardware and software support.

## Documentation
* Added setup and training guidance for supported MLU models, containers, launch configuration, and known limitations.
* Added MLU hardware support to the documentation navigation.

### CODE REVIEW

File: veomni/utils/helper.py [hunk 1]
```diff
@@ -728,6 +740,10 @@ def handler_fn(p):
             profiler_level=torch_npu.profiler.ProfilerLevel.Level1,
             data_simplification=False,
         )
+    elif IS_MLU_AVAILABLE:
+        profiler_module = torch.profiler
+        activities = [profiler_module.ProfilerActivity.CPU, profiler_module.ProfilerActivity.MLU]
```
[the remark this answers is not in the document]
Comment (sc1915) [reply veomni/utils/helper.py:787]: 此实现基于torch_mlu (IS_MLU_AVAILABLE), 确认无误

File: veomni/arguments/arguments_types.py [hunk 2]
```diff
@@ -986,6 +986,20 @@ def _resolve_profile(self):
...
+    "moe_implementation": frozenset({"fused_mlu", "fused_mlu_triton"}),
+}
+
+_MLU_DEFAULT_FALLBACK: Dict[str, str] = {
+    "rms_norm_implementation": "eager",
+    "rotary_pos_emb_implementation": "eager",
+    "swiglu_mlp_implementation": "eager",
+    "load_balancing_loss_implementation": "eager",
+    "cross_entropy_loss_implementation": "eager",
+    "moe_implementation": "fused_mlu",
```
Comment (FoolPlayer) [inline veomni/arguments/arguments_types.py]: Default always maps to `fused_mlu`, then validation raises if apex is missing. The `'fused'` alias already falls back to `fused_mlu_triton` without apex — please do the same here.

File: veomni/arguments/arguments_types.py [hunk 3]
```diff
@@ -986,6 +986,20 @@ def _resolve_profile(self):
     "moe_implementation": "fused_npu",
 }
 
+# MLU compatibility tables for ``_validate_implementations``.
+_MLU_ALLOWED: Dict[str, frozenset] = {
+    "moe_implementation": frozenset({"fused_mlu", "fused_mlu_triton"}),
```
Comment (FoolPlayer) [inline veomni/arguments/arguments_types.py:989]: `_MLU_ALLOWED` only covers MoE, unlike `_NPU_ALLOWED` (every registered op + coverage assert). User overrides such as `rms_norm=liger_kernel` skip parse-time checks. Please mirror the NPU tables.

File: veomni/utils/import_utils.py [hunk 4]
```diff
@@ -68,6 +70,16 @@ def is_liger_kernel_available() -> bool:
...
     return _PACKAGE_FLAGS["torch_npu"]
 
+def is_torch_mlu_available() -> bool:
+    return _PACKAGE_FLAGS["torch_mlu"]
+
+def is_apex_mlu_available() -> bool:
+    if not _PACKAGE_FLAGS["apex"]:
+        return False
+    try:
+        return "mlu" in importlib.metadata.version("apex")
```
Comment (FoolPlayer) [inline veomni/utils/import_utils.py]: Fragile: a Cambricon apex whose version is `0.1.0` (no `"mlu"` substring) is treated as missing. Prefer importing `apex.contrib.grouped_gemm` instead of matching the version string.

File: veomni/ops/kernels/moe/mlu_group_gemm.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+            hidden_states=hidden_states,
+            expert_mask=expert_mask,
+            num_experts=num_experts,
+            input_splits=input_splits,
+            output_splits=output_splits,
+            num_global_tokens_per_local_expert=num_global_tokens_per_local_expert,
+            ep_group=get_parallel_state().ep_group,
+        )
+
+        cumsum = torch.cumsum(num_global_sum_tokens_per_local_expert, dim=0)
```
Comment (FoolPlayer) [inline veomni/ops/kernels/moe/mlu_group_gemm.py]: `preprocess()` copies this tensor to CPU with `non_blocking=True`. GPU EP uses `dispatch_to_ep_class()`, which does `.to(permute_tokens.device)`. Please reuse that helper; if apex wants host `batch_sizes`, document it and don't race the async copy.

File: veomni/utils/device.py [hunk 6]
```diff
@@ -19,25 +19,28 @@
...
-from .import_utils import is_torch_npu_available
+from .import_utils import is_torch_mlu_available, is_torch_npu_available
 
 
 logger = logging.get_logger(__name__)
 
 
 IS_CUDA_AVAILABLE = torch.cuda.is_available()
 IS_NPU_AVAILABLE = is_torch_npu_available()
+IS_MLU_AVAILABLE = is_torch_mlu_available()
```
Comment (FoolPlayer) [inline veomni/utils/device.py:30]: `is_torch_mlu_available()` is a `find_spec` check only, and `torch_mlu` is never imported here. Please `import torch_mlu` and gate on `torch.mlu.is_available()`.

File: veomni/distributed/sequence_parallel/async_ulysses.py [hunk 7]
```diff
@@ -143,12 +147,13 @@ def forward(
...
             output_k, mean_k, invvar_k = None, None, None
-            if IS_CUDA_AVAILABLE:
+            if IS_CUDA_AVAILABLE or (IS_MLU_AVAILABLE and is_apex_mlu_available()):
+                # fused_layer_norm_cuda is supported in apex_mlu
                 global fused_layer_norm_cuda
                 if fused_layer_norm_cuda is None:
                     fused_layer_norm_cuda = importlib.import_module("fused_layer_norm_cuda")
             if norm_type == "rmsnorm":
-                if IS_CUDA_AVAILABLE:
+                if IS_CUDA_AVAILABLE or (IS_MLU_AVAILABLE and is_apex_mlu_available()):
```
Comment (FoolPlayer) [inline veomni/distributed/sequence_parallel/async_ulysses.py]: On MLU without apex the `else` still calls `torch_npu.npu_rms_norm`, but `torch_npu` is not imported → `NameError`. Add an MLU eager path or fail clearly; do not call NPU ops on MLU.

File: veomni/ops/kernels/moe/__init__.py [hunk 8]
```diff
@@ -102,12 +113,20 @@ def apply_veomni_fused_moe_patch(fused_moe_kernel: str = "triton") -> None:
         if is_torch_npu_available():
             raise RuntimeError("fused_moe_kernel='triton' is GPU-only. Use 'npu' on NPU devices.")
         if not is_fused_moe_available():
-            raise RuntimeError("fused_moe_kernel='triton' requires triton to be installed and a supported GPU.")
+            raise RuntimeError("fused_moe_kernel='triton' requires triton to be installed and a supported GPU or MLU.")
+        from .group_gemm import group_gemm_fused_moe_forward
+
+        _fused_moe_forward = group_gemm_fused_moe_forward
+    elif fused_moe_kernel == "mlu_triton":
```
Comment (FoolPlayer) [inline veomni/ops/kernels/moe/__init__.py:120]: `mlu_triton` only checks `torch_mlu`, unlike the `triton` branch which requires `is_fused_moe_available()`. Please check triton the same way.

File: veomni/ops/kernels/moe/_kernels/kernel/moe.py [hunk 9]
```diff
@@ -56,7 +56,7 @@ def expert_histogram(input: torch.Tensor, num_bins: int) -> torch.Tensor:
     probably should go for some other histogram method.
     """
 
-    assert input.is_cuda
+    assert (input.is_cuda or input.is_mlu)
```
Comment (FoolPlayer) [inline veomni/ops/kernels/moe/_kernels/kernel/moe.py]: Vanilla PyTorch tensors have no `is_mlu`. Use `tensor.device.type in ("cuda", "mlu")` or `getattr(tensor, "is_mlu", False)`. Same for the other asserts in this file.

File: veomni/trainer/base.py [hunk 10]
```diff
@@ -72,6 +72,7 @@
     get_dist_comm_backend,
     get_torch_device,
     is_nccl_backend,
+    is_cncl_backend,
```
Comment (FoolPlayer) [inline veomni/trainer/base.py]: Unused import (ruff F401). Either use `is_cncl_backend` in teardown or drop this import.

File: veomni/ops/kernels/moe/_kernels/utils/device.py [hunk 11]
```diff
@@ -16,11 +16,16 @@
...
 from ......utils.device import get_device_name
 
+from veomni.utils.device import IS_MLU_AVAILABLE
+
 
 @lru_cache
 def get_device_key() -> str:
     import torch
 
+    if IS_MLU_AVAILABLE:
```
Comment (FoolPlayer) [inline veomni/ops/kernels/moe/_kernels/utils/device.py:26]: Returns `"MLU"` whenever the torch_mlu **package** is installed, before CUDA capability — can mis-key Triton autotune on a GPU host with `torch_mlu` installed.

**Review by FoolPlayer — CHANGES_REQUESTED** [review]
Thanks for the MLU bring-up — the overall approach looks right. Inline comments cover the code issues; please address those before merge.

**PR title:** `[MLU][feat]: ...` will fail title CI (`MLU` is not an allowed module, and there must be a space before `feat`). Please use something like `[ops, dist] feat: Add Cambricon MLU support`.

**Correctness check (blocking):** a MLU-only loss plot is not enough to catch EP permute / cumsum bugs. Please run a **GPU vs MLU comparison on an MoE model with EP enabled**, e.g. Qwen3-VL-30B-A3B or Qwen3-MoE, same YAML / data / seqlen / micro-batch, `--train.accelerator.ep_size 8`:
- GPU: `fused_triton` or `fused_quack`
- MLU: `fused_mlu` (and `fused_mlu_triton` if apex is optional)

Please report overlaid loss for the first N steps, whether grads stay finite, and any kernel fallback differences.

File: veomni/arguments/arguments_types.py (hunk 3 above)
Comment (sc1915) [reply veomni/arguments/arguments_types.py:989]: only fused_mlu is supported now, we will support  rest ops soon. other ops will fallback to "eager" through _MLU_DEFAULT_FALLBACK, is it enough for op check? I wonder if there is someother places need check?

File: veomni/utils/import_utils.py [hunk 12]
```diff
@@ -43,6 +43,8 @@ def _get_package_version(name: str) -> "Version":
     "flash_attn": _is_package_available("flash_attn"),
     "liger_kernel": _is_package_available("liger_kernel"),
     "torch_npu": _is_package_available("torch_npu"),
+    "torch_mlu": _is_package_available("torch_mlu"),
+    "apex": _is_package_available("apex"),
```
Comment (FoolPlayer) [inline veomni/utils/import_utils.py:47]: How to install MLU apex?

**Review by FoolPlayer — APPROVED** [review]

**Comment by FoolPlayer** [comment]
Thanks for your PR, one more question like CUDA or NPU, we use uv manage the python env, can we also use uv on MLU? It seems all packages are prepared only on mlu image.

**Comment by sc1915** [comment]
> Thanks for your PR, one more question like CUDA or NPU, we use uv manage the python env, can we also use uv on MLU? It seems all packages are prepared only on mlu image.

yes,it's ok to use uv to manage the python env

File: veomni/utils/import_utils.py (hunk 12 above)
Comment (sc1915) [reply veomni/utils/import_utils.py:47]: it's need apex mlu source code or .whl to install

File: veomni/utils/device.py (hunk 6 above)
Comment (sc1915) [reply veomni/utils/device.py:30]: for torch_mlu is unused, I del `imoprt torch_mlul` here, and use `torch.mlu.is_available()` to get device information, and it's enough in mlu
